eravm-stable-interface/src/lib.rs

```diff
@@ -0,0 +1,136 @@
+use u256::{H160, U256};
+
+pub trait StateInterface {
+    fn read_register(&self, register: u8) -> (U256, bool);
+    fn set_register(&mut self, register: u8, value: U256, is_pointer: bool);
+
+    fn number_of_callframes(&self) -> usize;
+    /// zero is the current frame, one is the frame before that etc.
+    fn callframe(&mut self, n: usize) -> impl CallframeInterface + '_;
+
+    fn read_heap_byte(&self, heap: HeapId, index: u32) -> u8;
```

Comment:
Why is it possible to only read / write single bytes at a time? This doesn't look very efficient / ergonomic. IMO, could make more sense to have methods to copy a range of bytes from / to heap, i.e., something like

```rust
fn copy_from_heap(&self, id: HeapId, offset: u32, buffer: &mut [u8]);
fn copy_to_heap(&mut self, id: HeapId, offset: u32, buffer: &[u8]);
```

Reply:
The problem with your suggestion is that it has worse performance because it introduces unnecessary copying/allocation. An iterator-based interface might be more ergonomic, though it is also more complex.

Ergonomics and performance are secondary goals here, though. For instance, any sane user would want to know what instruction is currently being executed but that is not part of the interface because the interface is stable and minimal.

To get the currently executed instruction, the user must get the pc, read the code page and then decode that. That can't be done in this library because zkevm_opcode_defs' extensions aren't done in a backwards compatible way. 😢 Because of that complexity, I think another ergonomics layer is warranted.

Reply:
How would you provide that nice interface?

One possibility is making a better zkevm_opcode_defs. I already have code that rips the instruction encoding.
https://github.com/matter-labs/eravm_instruction_encoding